### PR TITLE
try docker release script and dockerfiles

### PR DESCRIPTION
**Pull request by 8bitsam**
Add dockerfiles for windows + linux, use [https://github.com/sickcodes/Docker-OSX](url) for macos. Also add a python script that does the actual job of building the releases on each system.

Made the PR so the code could be seen. Not sure if we want to use it / if anyone else has other code that might be better.

Some more info is in the `RELEASE.md` file added

### CODE REVIEW

**Comment by sbillinge** [comment]
Thanks @8bitsam this looks pretty cool as a start.   Let's check with @Sparks29032 and combine best things from different approaches.

I would like to separate the building of packages from the upload if possible, so the upload is handled by our conventional scripts.  From a UC/workflow point of view t would be nice if the workflow was to run the conventional scripts, but activate the docker stuff with a command-line arg that just somehow replaces the build part of activity?  sthg like that?

**Comment by 8bitsam** [comment]
I've added a flag for upload, `-u`. I'll also add functionality to use the other release script for uploading on the next commit.

**Comment by 8bitsam** [comment]
Now this should use `basic_release.sh` for the uploading. See `RELEASE.md` for the full usage info. I copied the steps from `basic_release.txt` to make it easier to follow.

**Comment by 8bitsam** [comment]
The build seems to work for windows and linux but mac still causing issues...

**Comment by 8bitsam** [comment]
The builds on all OS's work now, but I modified the `basic_release.sh` script to have an additional option to pull docker images and then upload those. This way, the old functionality remains and `docker_release.py` script should be able to create docker images, upload them, and then delete them after so storage isn't taken up if the `-u` flag is selected.

We could do a dry run without the upload flag, so that we do the upload/release and docker image removal manually.

File: basic_release.sh [hunk 1]
```diff
@@ -3,22 +3,30 @@
 # Inputs
 # $1 -- Directory to release
 # $2 -- Version number
+# $3 -- Optional flag -docker
+# $4 -- package_name (required if -docker is used)
 
-if [ -z "${1+x}" ] && [ -z "${2+x}" ]
+if [ -z "${1+x}" ] || [ -z "${2+x}" ] || { [ "$3" = "-docker" ] && [ -z "${4+x}" ]; }
```
Comment (sbillinge) [inline basic_release.sh:9]: can we make this `--docker` to make the syntax consistent with `argparse` (even though it is an arg for bash).  I am not sure if this is allowed or good practice, but the less cognitive overload for my brain the better.

**Review by sbillinge — COMMENTED** [review]
pls see comment.

I will review the instructions shortly.

Has this been tested?  Is there a good way to test it before Tuesday?  Is it easier if I merge it and we test it and revert if we need to?

**Comment by 8bitsam** [comment]
Now, when you run the script, it builds to a local directory `built-wheels` containing a directory with the package name + version that has the wheel files in it (i.e. if doing `diffpy.pdffit2`, will have `built-wheels/diffpy.pdffit2:3.12-linux/dist/'). The `dist/` directory contains the `.whl` file.

Also, windows docker containers don't exactly work on linux and mac due to kernel differences, so the windows build will be disabled on non-windows systems when the script is run (still works on windows systems). This way, if you want to build the wheels for all OS's, you will want to run the script on a windows machine.

The build for mac still isn't working, hoping to have that fixed soon.

**Comment by sbillinge** [comment]
> Now, when you run the script, it builds to a local directory `built-wheels` containing a directory with the package name + version that has the wheel files in it (i.e. if doing `diffpy.pdffit2`, will have `built-wheels/diffpy.pdffit2:3.12-linux/dist/`). The `dist/` directory contains the `.whl` file.
> 
> Also, windows docker containers don't exactly work on linux and mac due to kernel differences, so the windows build will be disabled on non-windows systems when the script is run (still works on windows systems). This way, if you want to build the wheels for all OS's, you will want to run the script on a windows machine.
> 
> The build for mac still isn't working, hoping to have that fixed soon.

Hi @8bitsam this was a huge effort, but we may not use it so you can slow roll for now.  We are also exploring using Github Actions to do the builds and then download the wheels, but either way, it is lower priority than getting things working on conda-forge.   Just a heads-up so that effort isn't wasted.   Not to say we won't pivot back to this approach in the future, so the knowledge we have gained (as well as the Docker containers) are highly useful.

**Comment by bobleesj** [comment]
@8bitsam Hi Sam, could we close this issue?

**Pull request closed without merging**